File: src/agent_sdk/conversation.py
from __future__ import annotations

import copy
from typing import AsyncIterator, Type
from uuid import UUID, uuid4

from agent_sdk.config import Config, get_tier_chain, load_config
from agent_sdk.fallback import execute_with_fallback
from agent_sdk.logging_ import ConversationLogger
from agent_sdk.providers.base import Provider
from agent_sdk.registry import get_provider, resolve_model
from agent_sdk.types import Message, ModelInfo, Response, StructuredResponse, T, Tier


# ##################################################################
# conversation
# multi-turn conversation that maintains message history and delegates
# to providers via the fallback engine. use as an async context manager.
#
# example:
#   async with agent.conversation("book-editor") as chat:
#       outline = await chat.say("Write an outline", schema=Outline)
#       chapter1 = await chat.say("Write chapter 1")
#       async for chunk in chat.stream("Write chapter 2"):
#           print(chunk, end="", flush=True)
class Conversation:

    # ##################################################################
    # init
    # set up a new conversation with optional name, tier, system prompt,
    # provider/model overrides, and config. the logger is created on enter.
    def __init__(
        self,
        name: str | None = None,
        *,
        tier: Tier = Tier.HIGH,
        system: str | None = None,
        provider: str | None = None,
        model: str | None = None,
        config: Config | None = None,
    ) -> None:
        self._name = name
        self._tier = tier
        self._system = system
        self._provider_name = provider
        self._model_id = model
        self._config = config
        self._history: list[Message] = []
        self._logger: ConversationLogger | None = None
        self._conversation_id: UUID | None = None
        # ##################################################################
        # pre-populate system message if provided
        # the system message is always first in the history list
        if system is not None:
            self._history.append(Message(role="system", content=system))

    # ##################################################################
    # aenter
    # create the conversation logger and assign a unique conversation id
    async def __aenter__(self) -> Conversation:
        self._conversation_id = uuid4()
        cfg = self._config or load_config()
        self._logger = ConversationLogger(
            self._conversation_id,
            name=self._name,
            tier=self._tier.value,
            provider=self._provider_name,
            model=self._model_id,
            log_base=cfg.logging.directory,
        )
        return self

    # ##################################################################
    # aexit
    # close the logger on context exit regardless of exception state
    async def __aexit__(self, *args) -> None:
        if self._logger is not None:
            self._logger.close()

    # ##################################################################
    # say
    # add a user message, call the provider via fallback, append the
    # assistant response to history, and return the response object.
    # pass schema to receive a StructuredResponse with a parsed pydantic model.
    # tier overrides the instance-level tier for this call only.
    async def say(
        self,
        content: str,
        *,
        schema: Type[T] | None = None,
        tier: Tier | None = None,
        timeout: float = 120.0,
    ) -> Response | StructuredResponse:
        effective_tier = tier if tier is not None else self._tier
        self._history.append(Message(role="user", content=content))
        messages = list(self._history)

        provider_entry, model_info = self._resolve_provider_model(effective_tier)
        providers_chain = self._build_providers_chain(effective_tier, provider_entry)

        async def execute_fn(chain_entry: str) -> Response | StructuredResponse:
            pname, mid = _split_entry(chain_entry)
            prov = _require_provider(pname)
            minfo = _require_model(pname, mid, effective_tier)
            return await prov.complete(
                messages,
                minfo,
                schema=schema,
                timeout=timeout,
            )

        result = await execute_with_fallback(
            effective_tier.value,
            providers_chain,
            execute_fn,
            is_conversation=True,
            config=self._config,
            logger=self._logger,
        )

        self._history.append(Message(role="assistant", content=result.text))
        return result

    # ##################################################################
    # stream
    # add a user message, stream chunks from the provider, collect the
    # full text, append it to history, and yield each chunk to the caller.
    # tier overrides the instance-level tier for this call only.
    async def stream(
        self,
        content: str,
        *,
        tier: Tier | None = None,
        timeout: float = 120.0,
    ) -> AsyncIterator[str]:
        effective_tier = tier if tier is not None else self._tier
        self._history.append(Message(role="user", content=content))
        messages = list(self._history)

        provider_entry, _ = self._resolve_provider_model(effective_tier)
        providers_chain = self._build_providers_chain(effective_tier, provider_entry)

        # resolve a concrete provider/model for streaming — use first in chain
        first_entry = providers_chain[0] if providers_chain else provider_entry or ""
        pname, mid = _split_entry(first_entry)
        prov = _require_provider(pname)
        minfo = _require_model(pname, mid, effective_tier)

        chunks: list[str] = []

        async def _generate() -> AsyncIterator[str]:
            async for chunk in prov.stream(messages, minfo, timeout=timeout):
                chunks.append(chunk)
                yield chunk

        async for chunk in _generate():
            yield chunk

        full_text = "".join(chunks)
        self._history.append(Message(role="assistant", content=full_text))

    # ##################################################################
    # history property
    # returns a snapshot of the current conversation history
    @property
    def history(self) -> list[Message]:
        return list(self._history)

    # ##################################################################
    # name property
    # returns the conversation name
    @property
    def name(self) -> str | None:
        return self._name

    # ##################################################################
    # tier property
    # returns the configured tier for this conversation
    @property
    def tier(self) -> Tier:
        return self._tier

    # ##################################################################
    # fork
    # create a new Conversation with a copy of the current history.
    # changes to the fork do not affect the original.
    def fork(self, name: str | None = None) -> Conversation:
        forked = Conversation(
            name=name,
            tier=self._tier,
            provider=self._provider_name,
            model=self._model_id,
            config=self._config,
        )
        # system is already baked into history — copy history directly,
        # skipping the __init__ system message injection
        forked._history = copy.deepcopy(self._history)
        return forked

    # ##################################################################
    # resolve provider model
    # find the explicit provider+model override or derive from tier chain
    def _resolve_provider_model(self, tier: Tier) -> tuple[str | None, ModelInfo | None]:
        if self._provider_name and self._model_id:
            entry = f"{self._provider_name}:{self._model_id}"
            minfo = resolve_model(self._provider_name, self._model_id, tier=tier)
            return entry, minfo
        return None, None

    # ##################################################################
    # build providers chain
    # produce an ordered list of provider:model entries for fallback.
    # if a single override entry is set it forms a one-element chain;
    # otherwise the tier config chain is used.
    def _build_providers_chain(self, tier: Tier, override_entry: str | None) -> list[str]:
        if override_entry is not None:
            return [override_entry]
        cfg = self._config or load_config()
        return get_tier_chain(tier, cfg)


# ##################################################################
# split entry
# split a "provider:model_id" string into its two parts.
# returns ("", "") for malformed entries.
def _split_entry(entry: str) -> tuple[str, str]:
    if ":" not in entry:
        return ("", "")
    provider_name, model_id = entry.split(":", 1)
    return provider_name, model_id


# ##################################################################
# require provider
# look up a provider by name and raise RuntimeError if not available.
def _require_provider(name: str) -> Provider:
    prov = get_provider(name)
    if prov is None:
        raise RuntimeError(f"Provider '{name}' is not available")
    return prov


# ##################################################################
# require model
# look up a ModelInfo by provider name and model id, raising RuntimeError
# if the model cannot be resolved.
def _require_model(provider_name: str, model_id: str, tier: Tier) -> ModelInfo:
    minfo = resolve_model(provider_name, model_id, tier=tier)
    if minfo is None:
        raise RuntimeError(f"Model '{provider_name}:{model_id}' could not be resolved")
    return minfo

File: go/conversation.go
package dazagentsdk

import (
	"context"
	"fmt"
	"strings"
	"sync"

	"github.com/google/uuid"
)

// ConversationOption configures a Conversation via the functional options
// pattern.
type ConversationOption func(*conversationOpts)

type conversationOpts struct {
	tier     Tier
	system   string
	provider string
	model    string
	config   *Config
}

// WithTier sets the default model tier for the conversation.
func WithTier(tier Tier) ConversationOption {
	return func(o *conversationOpts) { o.tier = tier }
}

// WithSystem sets the system prompt prepended to the conversation history.
func WithSystem(system string) ConversationOption {
	return func(o *conversationOpts) { o.system = system }
}

// WithProvider pins the conversation to a specific provider name.
func WithProvider(provider string) ConversationOption {
	return func(o *conversationOpts) { o.provider = provider }
}

// WithModel pins the conversation to a specific model ID.
func WithModel(model string) ConversationOption {
	return func(o *conversationOpts) { o.model = model }
}

// WithConfig overrides the default config for this conversation.
func WithConfig(cfg *Config) ConversationOption {
	return func(o *conversationOpts) { o.config = cfg }
}

// Conversation manages a multi-turn exchange with an AI provider. It
// maintains message history, uses fallback across providers, and logs
// all events to a ConversationLogger.
type Conversation struct {
	name           string
	tier           Tier
	system         string
	providerName   string
	modelID        string
	config         *Config
	history        []Message
	logger         *ConversationLogger
	conversationID uuid.UUID
	mu             sync.Mutex
}

// NewConversation creates a new conversation with the given name and options.
// The conversation logger is created immediately and assigned a unique ID.
func NewConversation(name string, opts ...ConversationOption) *Conversation {
	o := &conversationOpts{
		tier: TierHigh,
	}
	for _, fn := range opts {
		fn(o)
	}

	cfg := o.config
	if cfg == nil {
		var err error
		cfg, err = LoadConfig()
		if err != nil {
			cfg = &Config{}
			cfg.applyDefaults()
		}
	}

	convID := uuid.New()
	logger := NewConversationLogger(name, cfg, convID)

	c := &Conversation{
		name:           name,
		tier:           o.tier,
		system:         o.system,
		providerName:   o.provider,
		modelID:        o.model,
		config:         cfg,
		history:        nil,
		logger:         logger,
		conversationID: convID,
	}

	// Pre-populate system message if provided
	if o.system != "" {
		c.history = append(c.history, Message{Role: "system", Content: o.system})
	}

	return c
}

// SayOption configures a single Say call via the functional options pattern.
type SayOption func(*sayOpts)

type sayOpts struct {
	schema any
	tier   *Tier
}

// WithSaySchema sets the JSON schema for structured output on this Say call.
func WithSaySchema(schema any) SayOption {
	return func(o *sayOpts) { o.schema = schema }
}

// WithSayTier overrides the model tier for this specific Say call.
func WithSayTier(tier Tier) SayOption {
	return func(o *sayOpts) { o.tier = &tier }
}

// Say adds a user message to the history, sends all messages to the provider
// via the fallback engine, appends the assistant response to history, and
// returns the Response. Optional SayOption values allow per-call schema and
// tier overrides.
func (c *Conversation) Say(ctx context.Context, prompt string, opts ...SayOption) (*Response, error) {
	o := &sayOpts{}
	for _, fn := range opts {
		fn(o)
	}

	c.mu.Lock()
	c.history = append(c.history, Message{Role: "user", Content: prompt})
	messages := make([]Message, len(c.history))
	copy(messages, c.history)
	tier := c.tier
	if o.tier != nil {
		tier = *o.tier
	}
	providerName := c.providerName
	modelID := c.modelID
	cfg := c.config
	c.mu.Unlock()

	chain := c.buildChain(tier, providerName, modelID)

	completeOpts := CompleteOpts{
		Schema: o.schema,
	}

	executeFn := func(providerEntry string) (*Response, error) {
		pname, mid := splitEntry(providerEntry)
		prov := GetProvider(pname, cfg)
		if prov == nil {
			return nil, fmt.Errorf("provider '%s' is not available", pname)
		}
		minfo := ResolveModel(pname, mid, &tier, cfg)
		if minfo == nil {
			return nil, fmt.Errorf("model '%s:%s' could not be resolved", pname, mid)
		}
		return prov.Complete(ctx, messages, *minfo, completeOpts)
	}

	result, err := ExecuteWithFallback(ctx, string(tier), chain, executeFn, cfg, true)
	if err != nil {
		return nil, err
	}

	c.mu.Lock()
	c.history = append(c.history, Message{Role: "assistant", Content: result.Text})
	c.mu.Unlock()

	c.logger.LogEvent("turn_complete", map[string]any{
		"model":   result.ModelUsed.QualifiedName(),
		"turn_id": result.TurnID.String(),
	})

	return result, nil
}

// Stream adds a user message to the history, opens a streaming connection to
// the provider, and returns a channel of StreamChunks. The full response text
// is appended to history once the channel is drained.
func (c *Conversation) Stream(ctx context.Context, prompt string) (<-chan StreamChunk, error) {
	c.mu.Lock()
	c.history = append(c.history, Message{Role: "user", Content: prompt})
	messages := make([]Message, len(c.history))
	copy(messages, c.history)
	tier := c.tier
	providerName := c.providerName
	modelID := c.modelID
	cfg := c.config
	c.mu.Unlock()

	chain := c.buildChain(tier, providerName, modelID)

	// Resolve the first provider in the chain for streaming
	if len(chain) == 0 {
		return nil, NewAgentError("no providers in chain", ErrorNotAvailable, nil)
	}
	firstEntry := chain[0]
	pname, mid := splitEntry(firstEntry)
	prov := GetProvider(pname, cfg)
	if prov == nil {
		return nil, NewAgentError(
			fmt.Sprintf("provider '%s' is not available", pname),
			ErrorNotAvailable, nil,
		)
	}
	minfo := ResolveModel(pname, mid, &tier, cfg)
	if minfo == nil {
		return nil, NewAgentError(
			fmt.Sprintf("model '%s:%s' could not be resolved", pname, mid),
			ErrorNotAvailable, nil,
		)
	}

	srcCh, err := prov.Stream(ctx, messages, *minfo, StreamOpts{})
	if err != nil {
		return nil, err
	}

	outCh := make(chan StreamChunk)
	go func() {
		defer close(outCh)
		var fullText strings.Builder
		for chunk := range srcCh {
			if chunk.Err == nil {
				fullText.WriteString(chunk.Text)
			}
			outCh <- chunk
		}
		c.mu.Lock()
		c.history = append(c.history, Message{Role: "assistant", Content: fullText.String()})
		c.mu.Unlock()
	}()

	return outCh, nil
}

// Fork creates a new Conversation with a deep copy of the current history.
// The fork is independent: changes to one do not affect the other.
func (c *Conversation) Fork(name string) *Conversation {
	c.mu.Lock()
	historyCopy := make([]Message, len(c.history))
	copy(historyCopy, c.history)
	tier := c.tier
	providerName := c.providerName
	modelID := c.modelID
	cfg := c.config
	c.mu.Unlock()

	convID := uuid.New()
	logger := NewConversationLogger(name, cfg, convID)

	return &Conversation{
		name:           name,
		tier:           tier,
		providerName:   providerName,
		modelID:        modelID,
		config:         cfg,
		history:        historyCopy,
		logger:         logger,
		conversationID: convID,
	}
}

// History returns a copy of the current conversation history.
func (c *Conversation) History() []Message {
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([]Message, len(c.history))
	copy(result, c.history)
	return result
}

// Name returns the conversation name.
func (c *Conversation) Name() string {
	return c.name
}

// ConversationID returns the unique ID for this conversation.
func (c *Conversation) ConversationID() uuid.UUID {
	return c.conversationID
}

// Close writes a final conversation_end event and closes the logger.
func (c *Conversation) Close() {
	if c.logger != nil {
		c.logger.Close()
	}
}

// buildChain produces the ordered list of "provider:model" entries for
// fallback. If provider and model are both set, returns a single-element
// chain. Otherwise uses the tier config chain.
func (c *Conversation) buildChain(tier Tier, providerName, modelID string) []string {
	if providerName != "" && modelID != "" {
		return []string{providerName + ":" + modelID}
	}
	return GetTierChain(tier, c.config)
}

// splitEntry splits a "provider:model_id" string into its two parts.
func splitEntry(entry string) (string, string) {
	idx := strings.Index(entry, ":")
	if idx < 0 {
		return "", ""
	}
	return entry[:idx], entry[idx+1:]
}
